### ISSUE TITLE
aar dependency not working

### ISSUE THREAD

**Issue by matart15** [bytedeco/javacpp-presets#303]
Then i tried to make this app(https://github.com/CrazyOrr/FFmpegRecorder) into aar library and call it from myApplication.
This library uses ffmpeg as dependency.

structure is something like this

`JavaCV.jar -> FFmpegRecorder.aar -> myApplication`

But app can't find dependency of aar.

```
java.lang.UnsatisfiedLinkError: org.bytedeco.javacpp.avcodec
                                                     at java.lang.Class.classForName(Native Method)
                                                     at java.lang.Class.forName(Class.java:324)
                                                     at org.bytedeco.javacpp.Loader.load(Loader.java:413)
                                                     at org.bytedeco.javacpp.Loader.load(Loader.java:381)
                                                     at org.bytedeco.javacpp.avcodec$AVPacket.<clinit>(avcodec.java:1650)
                                                     at org.bytedeco.javacv.FFmpegFrameRecorder.<init>(FFmpegFrameRecorder.java:149)
                                                     at org.bytedeco.javacv.FFmpegFrameRecorder.<init>(FFmpegFrameRecorder.java:129)
                                                     at com.example.matar.testapplication.FFmpegRecordActivity.initRecorder(FFmpegRecordActivity.java:337)
                                                     at com.example.matar.testapplication.FFmpegRecordActivity.access$000(FFmpegRecordActivity.java:50)
                                                     at com.example.matar.testapplication.FFmpegRecordActivity$2.run(FFmpegRecordActivity.java:162)
```

I already tried this and `{transitive=true}`
http://stackoverflow.com/questions/34872382/manually-adding-aar-with-dependency-pom-iml-file

What am i doing wrong

This is my build.gradle to create aar file 

```
apply plugin: 'com.android.library'
apply plugin: 'maven-publish'
apply plugin: 'maven'
//apply plugin: 'com.android.application'

android {
    compileSdkVersion 23
    buildToolsVersion "23.0.0"

    defaultConfig {
        minSdkVersion 14
        targetSdkVersion 22
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

dependencies {
//    compile fileTree(dir: 'libs', include: ['*.jar'])
    testCompile 'junit:junit:4.12'
    compile 'com.android.support:appcompat-v7:23.3.0'
    compile 'com.jakewharton:butterknife:7.0.1'
//    compile files('libs/ffmpeg.jar')
//    compile files('libs/javacpp.jar')
//    compile files('libs/javacv.jar')

    compile group: 'org.bytedeco', name: 'javacv', version: '1.1'
    compile group: 'org.bytedeco', name: 'javacpp', version: '1.1'
    compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1', classifier: 'android-arm'
    compile group: 'org.bytedeco.javacpp-presets', name: 'ffmpeg', version: '2.8.1-1.1', classifier: 'android-arm'

}

uploadArchives {
    repositories{
        mavenDeployer {
            pom.groupId = 'com.example.matar'
            pom.artifactId = 'testapplication'
            pom.version = '0.0.1-SNAPSHOT'
            // Add other pom properties here if you want (developer details / licenses)
            repository(url: "file://${System.env.HOME}/.m2/repository/")
        }
    }
}
publishing {
    publications {
        maven(MavenPublication) {
            groupId 'com.example.matar' //You can either define these here or get them from project conf elsewhere
            artifactId 'testapplication'
            version '0.0.1-SNAPSHOT'
            artifact "$buildDir/outputs/aar/app-release.aar" //aar artifact you want to publish

            //generate pom nodes for dependencies
            pom.withXml {
                def dependenciesNode = asNode().appendNode('dependencies')
                configurations.compile.allDependencies.each { dependency ->
                    def dependencyNode = dependenciesNode.appendNode('dependency')
                    dependencyNode.appendNode('groupId', dependency.group)
                    dependencyNode.appendNode('artifactId', dependency.name)
                    dependencyNode.appendNode('version', dependency.version)
                }
            }
        }
    }

    //publish to filesystem repo
    repositories{
        maven {
            url "$buildDir/repo"
        }
    }
}
```

here is my application using this aar

```
package com.example.matar.myapplication;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Intent intent = new Intent(this, com.example.matar.testapplication.FFmpegRecordActivity.class);
        intent.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }
}
```

```
apply plugin: 'com.android.application'
//apply plugin: 'com.android.library'

android {
    compileSdkVersion 24
    buildToolsVersion "24.0.0"
    defaultConfig {
        applicationId "com.example.matar.myapplication"
        minSdkVersion 15
        targetSdkVersion 24
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}
allprojects {
    repositories {
        jcenter()
        flatDir {
            dirs 'libs'
        }
//        maven {
//            url "file://${System.env.HOME}/.m2/repository/"
//        }
//        maven {
//            url "/Users/matar/workplace/android/test/TestApplication/app/build/repo"
//        }
    }
}

dependencies {
//    compile fileTree(include: ['*.jar'], dir: 'libs')
    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
        exclude group: 'com.android.support', module: 'support-annotations'
    })
    compile 'com.android.support:appcompat-v7:24.2.1'
    testCompile 'junit:junit:4.12'
    compile(name:'app-release', ext:'aar'){ transitive = true }

    compile group: 'org.bytedeco', name: 'javacv', version: '1.1'
    compile group: 'org.bytedeco', name: 'javacpp', version: '1.1'
    compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.0.0-1.1', classifier: 'android-arm'
    compile group: 'org.bytedeco.javacpp-presets', name: 'ffmpeg', version: '2.8.1-1.1', classifier: 'android-arm'

//    compile ('com.example.matar:testapplication:0.0.1-SNAPSHOT@aar'){transitive=true}
}
```

---

I also tried copying `*.so` files too

My *.so files are in `app/src/main/jniLibs/armeabi` just like this.
https://github.com/CrazyOrr/FFmpegRecorder/tree/master/app/src/main/jniLibs/armeabi

and they work with this case
JavaCV.jar -> FFmpegRecorder as application

and does not work with this case
JavaCV.jar -> FFmpegRecorder.aar -> myApplication as application

I copied *.so files to `app/src/main/jniLibs/armeabi` directory of both `FFmpegRecorder.aar` and `myApplication`.

**Comment by saudet** [comment]
Looks like in the case of AAR files, we need to copy the `*.so` files in the `distribution` folder, as with this project:
https://github.com/googlesamples/android-ndk/tree/master/hello-libs/distribution/gperf/lib/armeabi

If you could add a wiki page explaining how to do that, it would be great!

**Comment by matart15** [comment]
@saudet Sorry I tried creating`distribution ` folder. but could not make it work.
I am not very famillar with android and java development.
I have no idea why distribution folder is needed and where to point it from.

project I trying to create aar is this. (https://github.com/matart15/FFmpegRecorder)
project I trying to use that aar is this (https://github.com/matart15/FFmpegRecorder_loader)

If you have time can you look into it. Please. 

I am pulling all my hairs here

**Comment by matart15** [comment]
I used this app to see copied libs.
https://play.google.com/store/apps/details?id=com.xh.nativelibsmonitor.app&hl=en
![screenshot_20161017-121132](https://cloud.githubusercontent.com/assets/17027219/19424334/38d59856-9463-11e6-8f50-2491416d2ea5.png)

if this app working right then *.so files seems copied.

Is it possible that *.so files are copied and not being referenced?

**Comment by saudet** [comment]
Yes it's possible:
https://github.com/bytedeco/javacpp-presets/wiki/The-UnsatisfiedLinkError-X-File-%28a-real-experience%29

**Comment by matart15** [comment]
I implemented ReLinker in my MainActivity onCreate(). Now i have this error
```
10-17 17:49:54.901 21576-21576/? I/art: Late-enabling -Xcheck:jni
10-17 17:49:54.924 21576-21582/? I/art: Debugger is no longer active
10-17 17:49:54.947 21576-21576/? I/InstantRun: Instant Run Runtime started. Android package is com.example.matar.myapplication, real application class is null.
10-17 17:49:56.283 21576-21576/com.example.matar.myapplication W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
10-17 17:49:56.431 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of avcodec...
10-17 17:49:56.437 21576-21576/com.example.matar.myapplication E/linker: /data/app/com.example.matar.myapplication-2/lib/arm/libavcodec.so: has text relocations
10-17 17:49:56.438 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Loading the library normally failed: java.lang.UnsatisfiedLinkError: dlopen failed: /data/app/com.example.matar.myapplication-2/lib/arm/libavcodec.so: has text relocations
                                                                            at java.lang.Runtime.loadLibrary(Runtime.java:372)
                                                                            at java.lang.System.loadLibrary(System.java:1076)
                                                                            at com.getkeepsafe.relinker.SystemLibraryLoader.loadLibrary(SystemLibraryLoader.java:24)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:163)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:92)
                                                                            at com.example.matar.myapplication.MainActivity.onCreate(MainActivity.java:23)
                                                                            at android.app.Activity.performCreate(Activity.java:6251)
                                                                            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
                                                                            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
                                                                            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                            at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                            at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                            at android.os.Looper.loop(Looper.java:148)
                                                                            at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
10-17 17:49:56.438 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) avcodec (null) was not loaded normally, re-linking...
10-17 17:49:56.447 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of swresample...
10-17 17:49:56.451 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) swresample (null) was loaded normally!
10-17 17:49:56.451 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of avutil...
10-17 17:49:56.451 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) avutil (null) was loaded normally!
10-17 17:49:56.451 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of m...
10-17 17:49:56.456 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) m (null) was loaded normally!
10-17 17:49:56.457 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of z...
10-17 17:49:56.457 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) z (null) was loaded normally!
10-17 17:49:56.457 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Beginning load of dl...
10-17 17:49:56.459 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Loading the library normally failed: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "__cxa_finalize" referenced by "/system/lib/libdl.so"...
                                                                            at java.lang.Runtime.loadLibrary(Runtime.java:372)
                                                                            at java.lang.System.loadLibrary(System.java:1076)
                                                                            at com.getkeepsafe.relinker.SystemLibraryLoader.loadLibrary(SystemLibraryLoader.java:24)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:163)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:92)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:189)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
                                                                            at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:92)
                                                                            at com.example.matar.myapplication.MainActivity.onCreate(MainActivity.java:23)
                                                                            at android.app.Activity.performCreate(Activity.java:6251)
                                                                            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
                                                                            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
                                                                            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                            at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                            at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                            at android.os.Looper.loop(Looper.java:148)
                                                                            at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
10-17 17:49:56.459 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) dl (null) was not loaded normally, re-linking...
10-17 17:49:56.463 21576-21576/com.example.matar.myapplication V/HODOR: (hold the door) Looking for lib/armeabi/libdl.so in APK...
10-17 17:49:56.464 21576-21576/com.example.matar.myapplication D/AndroidRuntime: Shutting down VM
10-17 17:49:56.465 21576-21576/com.example.matar.myapplication E/AndroidRuntime: FATAL EXCEPTION: main
                                                                                 Process: com.example.matar.myapplication, PID: 21576
                                                                                 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.matar.myapplication/com.example.matar.myapplication.MainActivity}: com.getkeepsafe.relinker.MissingLibraryException: lib/armeabi/libdl.so
                                                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
                                                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                                     at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                                     at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                     at android.os.Looper.loop(Looper.java:148)
                                                                                     at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                     at java.lang.reflect.Method.invoke(Native Method)
                                                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                                  Caused by: com.getkeepsafe.relinker.MissingLibraryException: lib/armeabi/libdl.so
                                                                                     at com.getkeepsafe.relinker.ApkLibraryInstaller.installLibrary(ApkLibraryInstaller.java:85)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:180)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:92)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibraryInternal(ReLinkerInstance.java:189)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:136)
                                                                                     at com.getkeepsafe.relinker.ReLinkerInstance.loadLibrary(ReLinkerInstance.java:92)
                                                                                     at com.example.matar.myapplication.MainActivity.onCreate(MainActivity.java:23)
                                                                                     at android.app.Activity.performCreate(Activity.java:6251)
                                                                                     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
                                                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
                                                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                     at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                     at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                     at android.os.Looper.loop(Looper.java:148) 
                                                                                     at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                     at java.lang.reflect.Method.invoke(Native Method) 
                                                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 

```

**Comment by matart15** [comment]
changed `targetSdkVersion` and `compileSdkVersion` to 22. but no luck

**Comment by saudet** [comment]
"has text relocations" -> that means you're using an old version of the presets. Please use the latest version.

**Comment by matart15** [comment]
@saudet Thank you. 
I it worked with `ReLinker `.
I downgraded to `compileSdkVersion 22` and `targetSdkVersion 22`

One final question. Does this mean my app will crush on android 6?

**Comment by saudet** [comment]
Is this still happening when using FFmpeg 3.4.1 released with JavaCPP 1.4?
http://search.maven.org/#search%7Cga%7C1%7Cbytedeco%20ffmpeg

**Comment by matart15** [comment]
@saudet Project is finished with`targetSdkVersion 22` and no longer updated. I can't confirm it now. So we can close this issue.
